### PR TITLE
Fixes VSTS Bug 968422: [Feedback] VS for Mac freezes on git revert

### PR DESCRIPTION
**Pull request by mkrueger**
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/968422

Fixed the exceptions from the log. However the real problem seems to be the sgen crash in mono.

### CODE REVIEW

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/FileUpdateEventHandler.cs [hunk 1]
```diff
@@ -35,6 +35,9 @@ public class FileUpdateEventInfo
 	{
 		public FileUpdateEventInfo (Repository repo, FilePath filePath, bool isDirectory)
 		{
+			if (FilePath.IsNullOrEmpty) {
+				LoggingService.LogInternalError ("FilePath was null or empty.", new Exception (nameof (FilePath)));
```
Comment (Therzok) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/FileUpdateEventHandler.cs]: This won't log a stacktrace.

Comment (Therzok) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/FileUpdateEventHandler.cs]: Just throw an exception ArgimentException

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/DocumentRegistry.cs [hunk 2]
```diff
@@ -164,7 +164,7 @@ static List<DocumentInfo> GetAllChangedFiles ()
 
 		class DocumentInfo : IDisposable
 		{
-			public Document Document { get; private set; }
```
Comment (sevoku) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/DocumentRegistry.cs]: unrelated formatting change

File: main/src/addins/ChangeLogAddIn/CommitDialogExtensionWidget.cs [hunk 3]
```diff
@@ -157,7 +157,7 @@ public override bool OnBeginCommit (ChangeSet changeSet)
 					}
 					if (!cset.ContainsFile (ce.File)) {
 						Task.Run (async () => {
-							if (!(await cset.Repository.GetVersionInfoAsync (ce.File)).IsVersioned)
+							if (cset.Repository.TryGetVersionInfo (ce.File, out var info) && info.IsVersioned)
```
Comment (sevoku) [inline main/src/addins/ChangeLogAddIn/CommitDialogExtensionWidget.cs]: Inverted logic?

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/PublishCommand.cs [hunk 4]
```diff
@@ -87,7 +87,7 @@ static void GetFiles (List<FilePath> files, WorkspaceObject entry)
 			if (!VersionControlService.CheckVersionControlInstalled ())
 				return false;
 
-			if (!(await vc.GetVersionInfoAsync (path)).IsVersioned && isDir)
+			if (vc.TryGetVersionInfo (path, out var info) && info.IsVersioned && isDir)
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/PublishCommand.cs]: Inverted logic?

Comment (mkrueger) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/PublishCommand.cs]: fixed.

File: main/src/addins/ChangeLogAddIn/CommitDialogExtensionWidget.cs (hunk 3 above)
Comment (mkrueger) [reply main/src/addins/ChangeLogAddIn/CommitDialogExtensionWidget.cs]: fixed.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/DocumentRegistry.cs (hunk 2 above)
Comment (mkrueger) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/DocumentRegistry.cs]: reverted - does that matter much ?
We should clean up such things over time IMO.

Comment (sevoku) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/DocumentRegistry.cs]: sure, but please in separate commits, so that those changes don't get mixed with the important ones.

**Comment by sevoku** [comment]
@monojenkins backport release-8.3

**Comment by sevoku** [comment]
@monojenkins backport release-8.3

**Comment by sevoku** [comment]
@mkrueger still crashing for me, also on branch switches

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs [hunk 5]
```diff
@@ -298,12 +298,15 @@ public bool TryGetVersionInfo (FilePath localPath, VersionInfoQueryFlags queryFl
 		public async Task<IReadOnlyList<VersionInfo>> GetVersionInfoAsync (IEnumerable<FilePath> paths, VersionInfoQueryFlags queryFlags = VersionInfoQueryFlags.None, CancellationToken cancellationToken = default)
 		{
 			if ((queryFlags & VersionInfoQueryFlags.IgnoreCache) != 0) {
-				// We shouldn't use IEnumerable because elements don't save property modifications.
-				var res = await OnGetVersionInfoAsync (paths, (queryFlags & VersionInfoQueryFlags.IncludeRemoteStatus) != 0, cancellationToken);
-				foreach (var vi in res)
-					if (!vi.IsInitialized) await vi.InitAsync (this, cancellationToken);
-				await infoCache.SetStatusAsync (res, cancellationToken: cancellationToken);
-				return res;
+				var res = await ExclusiveOperationFactory.StartNew (async delegate {
```
Comment (Therzok) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: Do we still use the concurrent factory anywhere?

**Comment by mkrueger** [comment]
Can't reproduce crashes OnGetVersionInfoAsync runs on ExclusiveOperationFactory. 

I'll just kill the ConcurrentOperationFactory.

**Comment by mkrueger** [comment]
@monojenkins backport release-8.3

**Comment by Therzok** [comment]
Some files were deleted and they shouldn't have been.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs [hunk 6]
```diff
@@ -377,6 +377,30 @@ public bool TryGetVersionInfo (IEnumerable<FilePath> paths, VersionInfoQueryFlag
 			return true;
 		}
 
+		protected Task<TResult> StartNewTask<TResult> (Func<TResult> function, CancellationToken cancellationToken = default)
+		{
+			if (Thread.CurrentThread == exclusiveThread) {
```
Comment (Therzok) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: There is no guarantee that a task will run on a specific thread.

Comment (Therzok) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: The exclusive scheduler should run inline if there is no current task executing though.

**Comment by mkrueger** [comment]
Ok seems we've bug somewhere. I'm sure I didn't delete any file…

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs (hunk 6 above)
Comment (mkrueger) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: It's just a dead lock prevention that the same task can't yield another exclusive thread.

Comment (Therzok) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: If the problem here is re-entrancy, we need a different solution.

Comment (mkrueger) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/Repository.cs]: I removed that from the PR. The test was locking because of a method that needed to run in UI thread.

No time to fix that.

We should move away from libgit asap - it's causing too much issues. It's not suited for threading :/.

**Comment by sevoku** [comment]
@monojenkins backport release-8.3

**Comment by mkrueger** [comment]
backported to https://github.com/mono/monodevelop/pull/8483
